src/CanvasContainer.tsx

```diff
@@ -2,171 +2,400 @@ import React, { CSSProperties, useEffect, useRef } from 'react';
-import { isWithPlatformMetaKey, isTextInputLikeElement } from './utils';
+import { AnyState } from './types';
 
-const dragModel = createModel(
+interface DragSession {
+  pointerId: number;
+  point: Point;
+}
+
+interface PointDelta {
+  x: number;
+  y: number;
+}
+
+const dragSessionModel = createModel(
   {
-    startPoint: { x: 0, y: 0 },
-    dragPoint: { x: 0, y: 0 },
-    dx: 0,
-    dy: 0,
-    embed: undefined as EmbedContext | undefined,
+    session: null as DragSession | null,
+    ref: null as React.MutableRefObject<HTMLElement> | null,
   },
   {
     events: {
-      LOCK: () => ({}),
-      RELEASE: () => ({}),
-      GRAB: (point: Point) => ({ point }),
-      DRAG: (point: Point) => ({ point }),
-      UNGRAB: () => ({}),
+      DRAG_SESSION_STARTED: ({ pointerId, point }: DragSession) => ({
+        pointerId,
+        point,
+      }),
+      DRAG_SESSION_STOPPED: () => ({}),
+      DRAG_POINT_MOVED: ({ point }: Pick<DragSession, 'point'>) => ({ point }),
     },
   },
 );
 
-const dragMachine = dragModel.createMachine({
-  initial: 'checking_embedded_mode',
-  states: {
-    checking_embedded_mode: {
-      always: [
-        {
-          target: 'in_embedded_mode',
-          cond: (ctx) => !!ctx.embed?.isEmbedded && !ctx.embed.pan,
+const dragSessionTracker = dragSessionModel.createMachine(
+  {
+    preserveActionOrder: true,
+    initial: 'idle',
+    states: {
+      idle: {
+        invoke: {
+          id: 'dragSessionStartedListener',
+          src:
+            ({ ref }) =>
+            (sendBack) => {
+              const node = ref!.current!;
+              const listener = (ev: PointerEvent) => {
+                const isLeftButton = ev.button === 0;
+                if (isLeftButton) {
+                  sendBack(
+                    dragSessionModel.events.DRAG_SESSION_STARTED({
+                      pointerId: ev.pointerId,
+                      point: {
+                        x: ev.pageX,
+                        y: ev.pageY,
+                      },
+                    }),
+                  );
+                }
+              };
+              node.addEventListener('pointerdown', listener);
+              return () => node.removeEventListener('pointerdown', listener);
+            },
         },
-        'released',
-      ],
-    },
-    in_embedded_mode: {},
-    released: {
-      meta: {
-        cursor: 'default',
-      },
-      invoke: {
-        src: 'invokeDetectLock',
-      },
-      on: {
-        LOCK: 'locked',
-      },
-    },
-    locked: {
-      initial: 'idle',
-      entry: ['disableTextSelection'],
-      exit: ['enableTextSelection'],
-      on: { RELEASE: 'released' },
-      invoke: {
-        src: 'invokeDetectRelease',
-      },
-      states: {
-        idle: {
-          meta: {
-            cursor: 'grab',
+        on: {
+          DRAG_SESSION_STARTED: {
+            target: 'active',
+            actions: actions.forwardTo(SpecialTargets.Parent),
```

Comment:
This is the first time I see `SpecialTargets` from XState. is this the same as `sendParent` the event?

Reply:
Yes, it's just that I've wanted to use that with `forwardTo`